### PR TITLE
Support Rails Engine routings

### PR DESCRIPTION
**Pull request by sachin21**
Rails Engines routing isn't supported yet in js_rails_routes but my project is using Rails Engine's routings so I added the function to export JavaScript routes for its Rails Engine.

### Support Rails Engines

- Add `exclude_engines` class variable
- Change a class variable name to base_path from path.
- Save the routes file each Rails Engines and `Rails.application`

I did these operation checkings.

- Install this changes with `gem 'js_rails_routes', path: '../../sachin21/js_rails_routes'`
- Execute `rails js:routes` in Rails project with Rails Engine routes.
- Confirmed it includes Rails Engine's routings.

### CODE REVIEW

**Comment by sachin21** [comment]
@yuku Hi, how about this?

File: spec/js_rails_routes/generator_spec.rb [hunk 1]
```diff
@@ -65,11 +65,11 @@
       end
 
       let(:include_names) do
-        /user/
+        /user|note/
```
Comment (yuku) [inline spec/js_rails_routes/generator_spec.rb]: Why this change is necessary?

File: lib/js_rails_routes/generator.rb [hunk 2]
```diff
@@ -57,15 +62,28 @@ def handle_route(route_name, route_path)
...
+      routes
+        .select(&:name)
+        .map { |r| [r.name, r.path.spec.to_s.split('(')[0]] }
+        .sort { |a, b| a[0] <=> b[0] }
     end
 
-    def write(string)
-      File.open(path, 'w') { |f| f.write(string) }
+    def write(klass, string)
+      File.open(File.join(base_path, "#{klass.gsub('::Engine', '').downcase}-routes.js"), 'w') { |f| f.write(string) }
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: `klass` tends to be used for an instance of `Class` class, but in this case it is a `String` here. Please rename it as `rails_engine_name` for readability.

File: lib/js_rails_routes/generator.rb [hunk 3]
```diff
@@ -57,15 +62,28 @@ def handle_route(route_name, route_path)
...
+      engines = Rails::Engine.subclasses.reject { |klass| klass.to_s.downcase =~ exclude_engines }
+
+      engine_routes = engines.map do |engine|
+        routes = make_routes(engine.routes.routes)
+        [engine.to_s, routes]
+      end
+      engine_routes = engine_routes.reject { |_, routes| routes.empty? }
+      routes = engine_routes.push(['Rails', make_routes(Rails.application.routes.routes)])
+
+      @routes ||= routes
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: `@name ||= name` is a idiom in Ruby to cache a result of heavy process. Whenever this method is called, the lines between 66 and 73 are executed. Put these lines in `begin` block to work as expected:

```rb
def routes_with_engines
  @routes_with_engines ||= begin
    # L66-L73 come here
  end
end
```

File: lib/js_rails_routes/generator.rb [hunk 4]
```diff
@@ -18,24 +18,29 @@ class Generator
...
     end
 
     def generate(task)
-      lines = ["// Don't edit manually. `rake #{task}` generates this file.", PROCESS_FUNC]
-      lines += routes.map do |route_name, route_path|
-        handle_route(route_name, route_path) if match?(route_name, route_path)
-      end.compact
-      lines += [''] # End with new line
-      write(lines.join("\n"))
+      routes_with_engines.each do |klass, routes|
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: Please  rename`klass` as `rails_engine_name`

File: lib/js_rails_routes/generator.rb [hunk 5]
```diff
@@ -57,15 +62,28 @@ def handle_route(route_name, route_path)
...
-                       .map { |r| [r.name, r.path.spec.to_s.split('(')[0]] }
-                       .sort { |a, b| a[0] <=> b[0] }
+    def routes_with_engines
+      engines = Rails::Engine.subclasses.reject { |klass| klass.to_s.downcase =~ exclude_engines }
+
+      engine_routes = engines.map do |engine|
+        routes = make_routes(engine.routes.routes)
+        [engine.to_s, routes]
+      end
+      engine_routes = engine_routes.reject { |_, routes| routes.empty? }
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: Please use method chain to avoid reassigning `engine_routes` variable:

```rb
engine_routes = engines
  .map { |engine| [engine.to_s. make_routes(engine.routes.routes)] }
  .reject { |_, routes| routes.empty? }
```

File: lib/js_rails_routes/generator.rb [hunk 6]
```diff
@@ -57,15 +62,28 @@ def handle_route(route_name, route_path)
...
+      routes = engine_routes.push(['Rails', make_routes(Rails.application.routes.routes)])
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: Since there is no need to mutate the `engine_routes`, please use `+` operator instead:

```rb
routes = engine_routes + ['Rails', make_routes(Rails.application.routes.routes)]
```

**Comment by yuku** [comment]
@sachin21 Hi, thanks for sending the PR. I can understand your motivation. Cloud you write tests?

File: spec/js_rails_routes/generator_spec.rb (hunk 1 above)
Comment (sachin21) [reply spec/js_rails_routes/generator_spec.rb]: Uhh, sorry. my bad.

File: lib/js_rails_routes/generator.rb (hunk 3 above)
Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: I see. got it.

**Comment by sachin21** [comment]
@yuku got it, I'll write tests :bowing_man:

File: lib/js_rails_routes/generator.rb (hunk 4 above)
Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: Make sense. it's a string.

File: lib/js_rails_routes/generator.rb (hunk 5 above)
Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: Good point.

File: lib/js_rails_routes/generator.rb (hunk 6 above)
Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: It does not return correct values because `engine_routes` is nested array.
So I wrote below.

```
ary + [['Rails', make_routes(Rails.application.routes.routes)]]
``

File: lib/js_rails_routes/generator.rb (hunk 2 above)
Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: Make a good point. :+1:

File: lib/js_rails_routes/generator.rb [hunk 7]
```diff
@@ -57,15 +62,27 @@ def handle_route(route_name, route_path)
...
+        .sort { |a, b| a[0] <=> b[0] }
     end
 
-    def write(string)
-      File.open(path, 'w') { |f| f.write(string) }
+    def write(rails_engine_name, string)
+      File.open(
+        File.join(base_path, "#{rails_engine_name.gsub('::Engine', '').downcase}-routes.js"),
+        'w'
+      ) { |f| f.write(string) }
```
Comment (yuku) [inline lib/js_rails_routes/generator.rb]: It's ok, but `IO.write` may improve readability IMO.

```rb
file_name = File.join(base_path, "#{rails_engine_name.gsub('::Engine', '').downcase}-routes.js")
File.write(file_name, string)
```

How do you think?

Comment (sachin21) [reply lib/js_rails_routes/generator.rb]: Agreed. I fixed that.

**Comment by yuku** [comment]
I'll publish next version after updating README

**Comment by sachin21** [comment]
@yuku Thank you so much! :+1:
